engine/googletest_helpers.cpp

```diff
@@ -0,0 +1,344 @@
+"[==========] 1 test from 1 test suite ran. (0 ms total)\n"
+"[  PASSED  ] 0 tests.\n"
+"[  FAILED  ] 1 test, listed below:\n"
+"[  FAILED  ] Suite.fail\n"
+"\n"
+" 1 FAILED TEST\n"
+);
+
+/// A test scenario that reports some tests as failed.
+static void
+test_fail(void)
+{
+    std::cout << fail_message2;
+    std::exit(EXIT_FAILURE);
+}
+
+
+/// An example pass message for `test_pass()`.
+const char pass_message2[] = (
+"Note: Google Test filter = Suite.pass\n"
+"[==========] Running 1 test from 1 test suite.\n"
+"[----------] Global test environment set-up.\n"
+"[----------] 1 test from Suite\n"
+"[ RUN      ] Suite.pass\n"
+"[       OK ] Suite.pass (0 ms)\n"
+"[----------] 1 test from Suite (0 ms total)\n"
+"\n"
+"[----------] Global test environment tear-down\n"
+"[==========] 1 test from 1 test suite ran. (0 ms total)\n"
+"[  PASSED  ] 1 test.\n"
+);
+
+/// A test scenario that passes.
+static void
+test_pass(void)
+{
+    std::cout << pass_message2;
+}
+
+
+/// An example pass message for `test_pass_but_exit_failure()`.
+const char pass_message3[] = (
+"Note: Google Test filter = Suite.pass_but_exit_failure\n"
+"[==========] Running 1 test from 1 test suite.\n"
+"[----------] Global test environment set-up.\n"
+"[----------] 1 test from Suite\n"
+"[ RUN      ] Suite.pass_but_exit_failure\n"
+"[       OK ] Suite.pass_but_exit_failure (0 ms)\n"
+"[----------] 1 test from Suite (0 ms total)\n"
+"\n"
+"[----------] Global test environment tear-down\n"
+"[==========] 1 test from 1 test suite ran. (0 ms total)\n"
+"[  PASSED  ] 1 test.\n"
+);
+
+/// A test scenario that passes but then exits with non-zero.
+static void
+test_pass_but_exit_failure(void)
+{
+    std::cout << pass_message3;
+    std::exit(70);
+}
+
+
+/// An example incomplete output for `test_timeout`.
+const char incomplete_test[] = (
+"Note: Google Test filter = Suite.incomplete\n"
+"[==========] Running 1 test from 1 test suite.\n"
+"[----------] Global test environment set-up.\n"
+"[----------] 1 test from Suite\n"
+"[ RUN      ] Suite.incomplete\n"
+);
+
+/// A test scenario that times out.
+///
+/// Note that the timeout is defined in the Kyuafile, as the TAP interface has
+/// no means for test programs to specify this by themselves.
+static void
+test_timeout(void)
+{
+    std::cout << incomplete_test;
+
+    ::sleep(10);
+    const fs::path control_dir = fs::path(utils::getenv("CONTROL_DIR").get());
+    std::ofstream file((control_dir / "cookie").c_str());
+    if (!file)
+        fail("Failed to create the control cookie");
+    file.close();
+}
+
+
+}  // anonymous namespace
+
+
+/// Prints out program usage and exits with a non-zero exit code.
+static void
+usage(const char* argv0) {
+    std::cout << "usage: " << argv0 << " "
+              << "[--gtest_color=*] "
+              << "[--gtest_filter=POSITIVE_PATTERNS] "
+              << "[--gtest_list_tests]"
+              << "\n\n"
+              << "This program mocks a googletest test program."
```

Comment:
Why mock a googletest program and not provide a real one, like `atf_helpers.cpp` does?

Reply:
> Why mock a googletest program and not provide a real one, like `atf_helpers.cpp` does?

The problem is that it would effectively require GoogleTest as a dependency. Are you ok with that in the longterm?

Reply:
In the long term, yes, I think I'm fine with a GoogleTest dependency. In fact... I have thought at times that it'd be nice to eliminate `atf-c++` altogether and rely on GoogleTest instead.

For now though, and given that this is used for just one test, the GoogleTest dependency could be made optional at configure time and the test or helpers built in that case only.

The point about having helpers for each variant of the output format seems good, but how likely is it going to be that the output changes anytime soon? Do you have a quote for the policy change you say post 1.9.0? I'm surprised they'll do that as I'm sure it'll break tons of things. Anyway, my real concern is that it seems hard to trust these tests because any subtle issue in what they print would make them unreliable.

Reply:
To be frank, we plan on embedding GoogleTest in Kyua, it doesn't make sense right now for the following reasons:
1. Upstream release doesn't support skipped tests, but we (FreeBSD) need to test skipped tests because of the fact that the latest release lacks this support >:/ (and it would make GoogleTest pretty dog gone useless for our purposes, otherwise, because it would be less featureful than ATF*)...
2. Pulling in master would be a mess with the autoconf integration (it usually lags, by a lot). cmake or bazel are the only really supported ways to integrate GoogleTest into Kyua (the former is community supported; the latter is Google supported).
3. I don't want to reinvent the wheel when it comes to pulling down Kyua and building its dependencies (there aren't any currently, but there might be more in the future).
4. About your question related to 1.9.0, I can't find the documentation I ran into before. It's changed since I last looked for it :/...

Can we please revisit this once 1.9.0 is released?